pkg/cmd/pulumi/templates/workspace.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templates
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+func (s *Source) getWorkspaceTemplates(
+	ctx context.Context, templateNamePathOrURL string, scope Scope, templateKind workspace.TemplateKind,
+	_ *sync.WaitGroup,
+) {
+	repo, err := workspace.RetrieveTemplates(ctx, templateNamePathOrURL, scope == ScopeLocal, templateKind)
+	if err != nil {
+		if notFound := new(workspace.TemplateNotFoundError); errors.As(err, notFound) {
+			s.addErrorOnEmpty(notFound)
+			return
+		}
+		// Bail on all errors unless its a 401 from a Pulumi Cloud backend...
+		if !errors.Is(err, workspace.ErrPulumiCloudUnauthorized) {
+			s.addError(err)
+			return
+		}
+
+		// ...If the request has 401'd AND we've identified the backend as being a Pulumi Cloud instance, we can
+		// attempt to retrieve the template using the user's Pulumi Cloud credentials.
+		repo, err = retrievePrivatePulumiCloudTemplate(templateNamePathOrURL)
+		if err != nil {
+			s.addError(err)
+			return
+		}
+	}
+	s.addCloser(repo.Delete)
+	workspaceTemplates, err := repo.Templates()
+	if err != nil {
+		s.addError(err)
+		return
+	}
+
+	s.addDownloadedTemplates(workspaceTemplates)
+}
+
+// Retrieve a Private template from the given Pulumi Cloud URL **including an auth token for Pulumi Cloud**.
+//
+// workspace.GetAccount ensures the user has a valid session with the Pulumi Cloud backend.
+//   - If the user is not logged in, the login flow will be initiated.
```

Comment:
GetAccount doesn't trigger login flow, it just reads from the credential file. All the login logic lives in pkg, if you need access to that here we're gonna have to do some significant refactoring.

Reply:
Ah, my bad there.

This comment was left over from me trying to understand this function (which was factored out from elsewhere). We don't need to refactor.